### ISSUE TITLE
Full MySQL test get IP blacklisted

### ISSUE THREAD

**Issue by sdann** [drwetter/testssl.sh#840]
MySQL will blacklist client IPs that make more than 100 bad TLS requests in a row. This is to protect it from nefarious scans. A full (no options) run of testssl.sh will go over this 100 connection limit in the Testing Vulnerabilities section of version 2.9.5.

The easy workaround is to set the following parameter in mysqld.conf and restart the server:

max_connect_errors = 1000

https://dev.mysql.com/doc/refman/5.7/en/server-system-variables.html#sysvar_max_connect_errors

> If more than this many successive connection requests from a host are interrupted without a successful connection, the server blocks that host from further connections. You can unblock blocked hosts by flushing the host cache. To do so, issue a FLUSH HOSTS statement or execute a mysqladmin flush-hosts command. If a connection is established successfully within fewer than max_connect_errors attempts after a previous connection was interrupted, the error count for the host is cleared to zero. However, once a host is blocked, flushing the host cache is the only way to unblock it. The default is 100.

However, I think we can be smart in the tool, and intersperse successful connections within the scan, to reset the counter. More investigation needed.

**Comment by drwetter** [comment]
Okay, keep me posted pls. Some ideas:

At recent pentest I discovered a similar thing. There it was an FTPS service which wasn't possible to scan with testssl.sh. nmap worked though. I suspect - but haven't looked at it - that this is due to the fact that testssl.sh tears down the whole connection, TLS and MySQL/FTP. Don't know whether it is possible just to leave TLS but leave the application protocol alive.

Alternatively we could do throttling. That shouldn't be difficult, e.g. ``OPENSSL="sleep 60; $OPENSSL``

**Comment by sdann** [comment]
I'm looking into this again this week after some other projects just finished. My guess is the STARTTLS protocol can't be restarted on an existing MySQL connection, but it's worth poking at to be sure.

**Comment by sdann** [comment]
Unfortunately, I can't find an easy way to reset the bad connection counter from a client, other than completing a valid authentication. There's indications that a failed login attempt will reset the counter (http://mysqlblog.fivefarmers.com/2013/08/08/understanding-max_connect_errors/) but I've tested this on the newest release (5.7.18) and it doesn't work.

The obvious path forward would be to have the administrator setup a test account, provide the credentials to testssl.sh, and make a few successful authentications during the test. However, doing this is about the same work for the MySQL administrator as just temporarily setting `max_connect_errors = 1000`. So I don't believe it's worth the effort to add to testssl.sh.

I'm going to look into the current, total, number of connections made, and see if it makes sense to split the MySQL test into a limited and full version.

**Comment by sdann** [comment]
A full test with 2.9.5 generates 171 connection errors, 71 more than the default 100 that MySQL is configured to start blocking at.

Skipping cipher ordering and client simulation tests will likely drop this below 100.

**Comment by drwetter** [comment]
that sounds promising to pursue. Did you try --fast?

2.9dev will have a bit more connections because of TLS1.3.
-- 
Sent from my mobile. Excuse my brevity&typos+the phone's autocorrection

**Comment by drwetter** [comment]
Steven, there was often "2,9.5" in this thread. Please note that this is a stable release and there won't be any new features. Only bugs will be backported. 

Other than that: all action takes place in 2.9dev. :-)

**Comment by drwetter** [comment]
**Edit:** bugfixes will be backported -)

**Comment by sdann** [comment]
Yep, will do. There's a bug in 2.9dev preventing --starttls=mysql from 
working. So I switched back to stable to focus on only this issue first.

**Comment by drwetter** [comment]
That would be great to get that fixed.

Can you do that or is that possible to provide enough information so that others can fix it?

**Comment by sdann** [comment]
Yes, I've debugged that issue and opened #914. I'm working on a fix.

Using `--fast` drops the total connection count from 171 to 139. Better, but still too many. Profiling each subsection, `-S` makes 48 connections, which seems like far too many for a set of tests that mostly dissects the server certificate.

The connection counts on all sections are:

| Section | CLI Arg | # Connections in 2.9.5 |
| --- | --- | --- |
| All | | 171 |
| Protocols | -p | 8 |
| Standard Cipher Categories | -s | 12 |
| PFS | -f | 7 |
| Server Preference | -P | 42 |
| Server Defaults | -S | 48 |
| Vulnerabilities | -U | 47 |
| Ciphers | -e | 18 |
| Client Simulation | -c | 7 |

**Comment by drwetter** [comment]
Oh, thanks for the statistics, @sdann !

As for the # of handshakes: As said in 2.9dev it will increase more or less because of TLS 1.3. If the server side doesn't support TLS 1.3 it shouldn't be much. As a guess I would say you're probably around 145 connections with --fast and no TLS 1.3 then. Getting below 100 sounds to me quite a task.

Admittedly the  Server Defaults sounds much.
